src/unix/mmap.c

```diff
@@ -345,14 +345,22 @@ closure_function(3, 3, boolean, mincore_fill_vec,
     u64 pgoff, i;
 
     if (pt_entry_is_present(e)) {
-        pgoff = (addr - bound(base)) >> PAGELOG;
-
         if (pt_entry_is_fat(level, e)) {
-            /* whole level is mapped */
-            for (i = 0; (i < 512) && (pgoff + i < bound(nr_pgs)); i++) {
-                bound(vec)[pgoff + i] = 1;
-	    }
+            /* have to handle starting in the middle of fat page */
+            u64 off, coff;
+            off = (addr & PAGEMASK_2M) >> PAGELOG;
+            if (addr > bound(base)) {
+                pgoff = (addr - bound(base)) >> PAGELOG;
+                coff = 0;
+            } else {
+                pgoff = 0;
+                coff = off;
+            }
+            for (i = 0; (i + coff < 512) && (pgoff + off + i < bound(nr_pgs)); i++) {
```

Comment:
Just a nit, but I think some expressions here could be simplified if i started at pgoff.

Reply:
It always fails in the stack. I'm not sure about making a test for it, because it's not always that the stack is in a 2M page.